File: ios/CareLog/CareLog/FHIR/Models/FHIRDocumentReference.swift
//
//  FHIRDocumentReference.swift
//  CareLog
//
//  FHIR DocumentReference resource model
//

import Foundation

/// CareLog representation of a FHIR DocumentReference resource.
struct FHIRDocumentReference: Codable, Identifiable {
    var id: String?
    let patientId: String
    let documentId: String  // UUID
    let type: DocumentType
    let title: String
    var description: String?
    let contentUrl: String  // S3 URL or presigned URL
    let contentType: String  // MIME type
    var size: Int64?
    let date: Date
    var authorId: String?
    var authorName: String?
    var authorType: PerformerType
    var status: DocumentStatus
    var periodStart: String?
    var periodEnd: String?

    init(
        id: String? = nil,
        patientId: String,
        documentId: String,
        type: DocumentType,
        title: String,
        description: String? = nil,
        contentUrl: String,
        contentType: String,
        size: Int64? = nil,
        date: Date,
        authorId: String? = nil,
        authorName: String? = nil,
        authorType: PerformerType = .relatedPerson,
        status: DocumentStatus = .current,
        periodStart: String? = nil,
        periodEnd: String? = nil
    ) {
        self.id = id
        self.patientId = patientId
        self.documentId = documentId
        self.type = type
        self.title = title
        self.description = description
        self.contentUrl = contentUrl
        self.contentType = contentType
        self.size = size
        self.date = date
        self.authorId = authorId
        self.authorName = authorName
        self.authorType = authorType
        self.status = status
        self.periodStart = periodStart
        self.periodEnd = periodEnd
    }
}

/// Document status as defined in FHIR.
enum DocumentStatus: String, Codable {
    case current
    case superseded
    case enteredInError = "entered-in-error"
}

// MARK: - FHIR JSON Conversion

extension FHIRDocumentReference {
    private static let iso8601Formatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    /// Convert to FHIR JSON representation.
    func toFHIRJSON() -> [String: Any] {
        var resource: [String: Any] = [
            "resourceType": "DocumentReference",
            "identifier": [
                [
                    "system": "https://carelog.com/document-id",
                    "value": documentId
                ]
            ],
            "status": status.rawValue,
            "type": [
                "coding": [
                    [
                        "system": "https://carelog.com/fhir/CodeSystem/document-type",
                        "code": type.rawValue,
                        "display": type.displayName
                    ],
                    [
                        "system": "http://loinc.org",
                        "code": type.loincCode,
                        "display": type.displayName
                    ]
                ]
            ],
            "category": [
                [
                    "coding": [
                        [
                            "system": "http://hl7.org/fhir/us/core/CodeSystem/us-core-documentreference-category",
                            "code": "clinical-note",
                            "display": "Clinical Note"
                        ]
                    ]
                ]
            ],
            "subject": ["reference": "Patient/\(patientId)"],
            "date": Self.iso8601Formatter.string(from: date)
        ]

        if let id = id {
            resource["id"] = id
        }

        if let description = description {
            resource["description"] = description
        }

        // Content/attachment
        var attachment: [String: Any] = [
            "contentType": contentType,
            "url": contentUrl,
            "title": title,
            "creation": Self.iso8601Formatter.string(from: date)
        ]
        if let size = size {
            attachment["size"] = size
        }
        resource["content"] = [["attachment": attachment]]

        // Author
        if authorId != nil || authorName != nil {
            var author: [String: Any] = [:]
            if let authorId = authorId {
                author["reference"] = "\(authorType.rawValue)/\(authorId)"
            }
            if let authorName = authorName {
                author["display"] = authorName
            }
            resource["author"] = [author]
        }

        // Context/period
        if periodStart != nil || periodEnd != nil {
            var period: [String: Any] = [:]
            if let start = periodStart { period["start"] = start }
            if let end = periodEnd { period["end"] = end }
            resource["context"] = ["period": period]
        }

        return resource
    }

    /// Parse from FHIR JSON.
    static func fromFHIRJSON(_ json: [String: Any]) throws -> FHIRDocumentReference {
        let id = json["id"] as? String

        guard let identifiers = json["identifier"] as? [[String: Any]],
              let documentId = identifiers.first?["value"] as? String else {
            throw FHIRClientError.decodingError("Missing document identifier")
        }

        guard let subject = json["subject"] as? [String: Any],
              let reference = subject["reference"] as? String else {
            throw FHIRClientError.decodingError("Missing subject reference")
        }
        let patientId = reference.replacingOccurrences(of: "Patient/", with: "")

        let typeDict = json["type"] as? [String: Any]
        let typeCoding = typeDict?["coding"] as? [[String: Any]]
        let typeCode = typeCoding?.first?["code"] as? String ?? "other"
        let type = DocumentType(rawValue: typeCode) ?? .other

        let statusString = json["status"] as? String ?? "current"
        let status = DocumentStatus(rawValue: statusString) ?? .current

        let description = json["description"] as? String

        guard let dateString = json["date"] as? String,
              let date = iso8601Formatter.date(from: dateString) else {
            throw FHIRClientError.decodingError("Missing or invalid date")
        }

        guard let content = json["content"] as? [[String: Any]],
              let attachment = content.first?["attachment"] as? [String: Any] else {
            throw FHIRClientError.decodingError("Missing content attachment")
        }

        let contentUrl = attachment["url"] as? String ?? ""
        let contentType = attachment["contentType"] as? String ?? ""
        let title = attachment["title"] as? String ?? ""
        let size = attachment["size"] as? Int64

        return FHIRDocumentReference(
            id: id,
            patientId: patientId,
            documentId: documentId,
            type: type,
            title: title,
            description: description,
            contentUrl: contentUrl,
            contentType: contentType,
            size: size,
            date: date,
            status: status
        )
    }
}

File: android/app/src/main/java/com/carelog/fhir/models/FhirDocumentReference.kt
package com.carelog.fhir.models

import com.carelog.fhir.client.DocumentType
import kotlinx.serialization.Serializable

/**
 * CareLog representation of a FHIR DocumentReference resource.
 *
 * Used for linking uploaded documents (prescriptions, reports) to FHIR.
 */
@Serializable
data class FhirDocumentReference(
    val id: String? = null,
    val patientId: String,
    val documentId: String,  // UUID
    val type: DocumentType,
    val title: String,
    val description: String? = null,
    val contentUrl: String,  // S3 URL or presigned URL
    val contentType: String,  // MIME type
    val size: Long? = null,
    val date: String,  // ISO 8601 datetime
    val authorId: String? = null,
    val authorName: String? = null,
    val authorType: PerformerType = PerformerType.RELATED_PERSON,
    val status: DocumentStatus = DocumentStatus.CURRENT,
    val periodStart: String? = null,
    val periodEnd: String? = null,
    val meta: ResourceMeta? = null
)

/**
 * Document status as defined in FHIR.
 */
@Serializable
enum class DocumentStatus(val fhirCode: String) {
    CURRENT("current"),
    SUPERSEDED("superseded"),
    ENTERED_IN_ERROR("entered-in-error")
}

/**
 * Convert to FHIR JSON format.
 */
fun FhirDocumentReference.toFhirJson(): Map<String, Any> {
    val resource = mutableMapOf<String, Any>(
        "resourceType" to "DocumentReference",
        "identifier" to listOf(
            mapOf(
                "system" to "https://carelog.com/document-id",
                "value" to documentId
            )
        ),
        "status" to status.fhirCode,
        "type" to mapOf(
            "coding" to listOf(
                mapOf(
                    "system" to "https://carelog.com/fhir/CodeSystem/document-type",
                    "code" to type.code,
                    "display" to type.displayName
                ),
                mapOf(
                    "system" to "http://loinc.org",
                    "code" to type.loincCode,
                    "display" to type.displayName
                )
            )
        ),
        "category" to listOf(
            mapOf(
                "coding" to listOf(
                    mapOf(
                        "system" to "http://hl7.org/fhir/us/core/CodeSystem/us-core-documentreference-category",
                        "code" to "clinical-note",
                        "display" to "Clinical Note"
                    )
                )
            )
        ),
        "subject" to mapOf("reference" to "Patient/$patientId"),
        "date" to date,
        "content" to listOf(
            mapOf(
                "attachment" to buildMap {
                    put("contentType", contentType)
                    put("url", contentUrl)
                    put("title", title)
                    put("creation", date)
                    size?.let { put("size", it) }
                }
            )
        )
    )

    id?.let { resource["id"] = it }
    description?.let { resource["description"] = it }

    if (authorId != null || authorName != null) {
        resource["author"] = listOf(
            buildMap {
                authorId?.let { put("reference", "${authorType.fhirReference}/$it") }
                authorName?.let { put("display", it) }
            }
        )
    }

    if (periodStart != null || periodEnd != null) {
        resource["context"] = mapOf(
            "period" to buildMap {
                periodStart?.let { put("start", it) }
                periodEnd?.let { put("end", it) }
            }
        )
    }

    return resource
}
